enhancements/authentication/invalid-certs.md

```diff
@@ -0,0 +1,200 @@
+Verification of server certificates is executed during TLS client handshakes,
+a TLS (https) client observing an invalid certificate will reject the connection attempt.
+
+Cluster internal issued certificates are not affected,
+however custom certificates can be configured in various cases:
+- custom serving certificates for kube-apiserver
+- custom API webhooks
+- custom aggregated API endpoints
+- custom certificates for route endpoints
+- certificates of external auth identity providers
+
+This will lead to broken connections to critical core parts of OpenShift and thus to a degraded cluster if invalid custom certificates are configured.
+
+This enhancement suggests a solution to prevent broken workloads from OpenShift 4.9 to OpenShift 4.10
+if invalid certificates are detected. It therefore proposes:
+- a way to detect invalid certificates in kube-apiserver and oauth-server
+- a way to prevent upgrades from OpenShift 4.9 to OpenShift 4.10 in the face of invalid certificates
+
+## Motivation
+
+Broken workloads must be prevented due to invalid certificates
+when upgrading from OpenShift 4.9 to OpenShift 4.10.
+
+### Goals
+
+A means to detect invalid certificates for TLS connections between:
+- kube-apiserver and API webhooks
+- kube-apiserver and aggregated API endpoints
+- custom serving certificates for kube-apiserver
+- oauth-server and its external route
+- oauth-server and external identity providers
+
+### Non-Goals
+
+Other core workloads are out of scope for this enhancement.
+
+## Proposal
+
+### User Stories
+
+1. As a cluster admin I want to be prevented from upgrading OpenShift 4.9 to OpenShift 4.10
+if I configured invalid certificates.
+
+2. As a cluster admin I want to get feedback in OpenShift 4.9 if invalid certificates are being detected.
+
+### API Extensions
+
+N/A
+
+### Detecting invalid certificates
+
+#### kube-apiserver
+
+In Kubernetes detection of invalid certificates has been added as part of https://github.com/kubernetes/kubernetes/pull/95396.
+The introduced Prometheus `apiserver_webhooks_x509_missing_san_total` and `apiserver_kube_aggregator_x509_missing_san_total` metrics provides the means to detect invalid certificates against API webhooks and aggregated API endpoints.
+
+To detect invalid certificates this enhancement proposes to add a new controller in `cluster-kube-apiserver` that:
+
+1. Regularly executes queries against the cluster internal Thanos instance
+2. Verifies if above metrics have a value >0
+3. If invalid certificates are detected, the controller sets a `InvalidCertsUpgradeable=false` condition on the `cluster` `KubeAPIServer.operator.openshift.io` resource which will yield the `Upgradeable=false` status on the `kube-apiserver` `config.openshift.io.ClusterOperator` resource.
+
+#### oauth-server
+
+For oauth-server this enhancement proposes the introduction of the following new metric:
+
+- `openshift_auth_external_x509_missing_san_total`: a metric capturing the count of invalid certificates
```

Comment:
What clients of this external server do we have?

Reply:
> What clients of this external server do we have?
If it is just the operator, could the operator build a custom verifier for the check?

Reply:
> If it is just the operator, could the operator build a custom verifier for the check?

The client, in this case, is oauth-server, not authentication-operator.